Implementation of the new API

vmx opened the pull request:
This is a huge PR. I've tried to keep the commits self-contained. So I suggest that your review one commit after another instead of the full PR. I'm happy to change the individual commits as I think for so many changes and breaking APIs it's important to have a clean history.

It might even make sense to review the "users" of that new API first, to get a feeling how this API really feels like:

 - https://github.com/ipfs/js-ipfs-unixfs-importer/pull/20
 - https://github.com/ipfs/js-ipfs-unixfs-exporter/pull/15
 - https://github.com/ipfs/js-ipfs-mfs/pull/35
 - https://github.com/ipfs/js-ipfs/pull/1848

vmx commented on src/index.js:
```diff
@@ -29,361 +25,450 @@ class IPLDResolver {
+    if (options.loadFormat === undefined) {
+      this.loadFormat = async (codec) => {
+        const codecName = multicodec.print[codec]
+        throw new Error(`No resolver found for codec "${codecName}"`)
```
`loadFormat` is only called when the format is not available. So I think it makes sense to throw if no function is given.

vmx commented on src/index.js:
```diff
@@ -29,361 +25,450 @@ class IPLDResolver {
+   */
+  put (nodes, userOptions) {
+    if (!typical.isIterable(nodes) || typical.isString(nodes) ||
+        Buffer.isBuffer(nodes)) {
```
I found having a `length` field too broad (which is what `isArrayLike`) is checking for.

achingbrain replied:
You can define a more specific check if you prefer, but being able to pass a [`bl`](https://www.npmjs.com/package/bl) instances around our stack will allow us to [increase performance elsewhere](https://github.com/ipfs/js-ipfs-unixfs-importer/pull/12#issuecomment-449437692).

`bl` instances fail the `Buffer.isBuffer` check so its use prevents this.

vmx replied:
I lean towards dealing with it once we get there :) Feel free to open an issue when it's released.

achingbrain replied:
Ok, sounds fine.

vmx commented on src/index.js:
```diff
@@ -29,361 +25,450 @@ class IPLDResolver {
+   */
+  remove (cids) {
+    if (!typical.isIterable(cids) || typical.isString(cids) ||
+        Buffer.isBuffer(cids)) {
```
I found having a `length` field too broad (which is what `isArrayLike`) is checking for.

vmx commented on src/index.js:
```diff
@@ -29,361 +25,450 @@ class IPLDResolver {
     waterfall([
-      (cb) => this._getFormat(cid.codec, cb),
+      async () => {
+        return this._getFormat(cid.codec)
```
The `async` module supports async/await style coding.

achingbrain replied:
Apparently this only works with non-transpiled async functions? "Using ES2017 async functions" -> https://caolan.github.io/async/ (can't deep link, sorry)

vmx replied:
Good catch. Does this mean that tests pass as we run webpack with different configurations for development and production? I even think we don't really run tests with the Browser build we publish.

achingbrain replied:
According to [Moz](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function#Browser_compatibility) async functions are supported everywhere except IE, and our contributing guidelines say [which Node version we support](https://github.com/ipfs/community/blob/master/CONTRIBUTING_JS.md#supported-versions) but not which browser types/versions.

Aegir uses [`babel/preset-env`](https://babeljs.io/docs/en/next/babel-preset-env.html) which I guess means it'll use native async unless you use IE, which will get a polyfill and this will break.

@alanshaw @olizilla @lidel - what browsers we officially support?

vmx replied:
For the browser build we currently use [this browser list](https://github.com/ipfs/aegir/blob/a5c1c32798579cc30e89f5c56089e8f0e16ef4df/src/utils.js#L30-L35):

[>1%, last 2 versions, Firefox ESR, not ie < 11](https://browserl.ist/?q=>1%25%2C+last+2+versions%2C+Firefox+ESR%2C+not+ie+<+11)

Which contains mobile browsers that don't support async/await.

achingbrain commented on test/ipld-bitcoin.js:
```diff
@@ -83,16 +83,12 @@ module.exports = (repo) => {
+      async function store () {
+        const nodes = [node1, node2, node3]
+        const result = resolver.put(nodes, multicodec.BITCOIN_BLOCK)
+        ;[cid1, cid2, cid3] = await result.all()
```
The leading semicolon makes me sad. I know why it's necessary, but it still makes me sad.

What about changing the variable scope and keeping everything contained within the function so you don't need it?  Yum yum, side-effect free.

vmx replied:
The tests really need a revisit at one point. It's a lot of duplicated code. I wanted to keep the changes as minimal as possible for this PR.

achingbrain replied:
Ha!

alanshaw commented on src/util.js:
```diff
@@ -0,0 +1,32 @@
+  iterator.last = () => exports.last(iterator)
+  iterator.all = () => exports.all(iterator)
+  return iterator
+}
```
The reason this is a decorator in IPFSX is because [generators have `return` and `throw` functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator) that are called automatically by a `for await` loop to allow the generator to clean up when execution is aborted/finished.

```js
const iterator = {
  return () {
    console.log('return called')
    return { done: true }
  },
  next () {
    return { done: false, value: 1 }
  },
  [Symbol.iterator] () {
    return iterator
  }
}

async function main () {
  let i = 0
  for await (const value of iterator) {
    i++
    if (i === 3) break
    console.log(value)
  }
}

main()

// output:
// 1
// 1
// return called
```

If we create a custom async iterator here that isn't a generator then breaking out of a loop that uses it will not call return, all that will happen is that `next` will not be called again. I don't think this is a problem _right now_ with the code in this PR...but there could be issues here e.g. `fs.createReadStream` with a transform that converts raw data into an IPLD node, passed to `ipld.put`, halting half way would never end the stream, could cause the whole of the file content to be buffered into memory.

vmx replied:
With "being a decorator" you mean that you're extending/modifying the iterator you passed in? I previously had that until @achingbrain mentioned that it's bad style to modify things you pass in (I agree with that). But in this case it might make sense.

What if fancyiterator just implements a return and throw? I know it won't be general purpose but good enough for js-ipld, won't it?

achingbrain replied:
Typically a [decorator](https://en.wikipedia.org/wiki/Decorator_pattern) would wrap the instance being decorated instead of modifying it, but you could argue it's a side-effect of a type system that doesn't allow modification at runtime.

vmx commented on src/index.js:
```diff
@@ -29,361 +25,450 @@ class IPLDResolver {
+        // get block
+        // use local resolver
+        // update path value
+        this.bs.get(cid, (err, block) => {
```
"Promisifying" things that way doesn't work. Here's a minimal example:

```javascript
'use strict'

const promisify = require('util').promisify

const CID = require('cids')
const IpfsBlockService = require('ipfs-block-service')
const IpfsRepo = require('ipfs-repo')

const ipfsRepoPath = '/tmp/getpromisifiedblocks'
const repo = new IpfsRepo(ipfsRepoPath, { init: true })

repo.init({}, (error) => {
  if (error) {
    throw error
  }

  repo.open(async (error) => {
    if (error) {
      throw error
    }
    const blockService = new IpfsBlockService(repo)

    const cid = new CID('zdpuAvwqFGiEVdAQjymUUyJzJK5nuZwn2QKw3WSj1vDdoZmRX')
    const block = await promisify(blockService.get)(cid)
    // Works (well it throws the error we would expect
    //blockService.get(cid, (error, block) => {
    //  if (error) {
    //    throw error
    //  }
    //})
  })
})
```

vmx replied:
If I would've had @alanshaw's sk1llz I would've known to just use `const block = await promisify(blockService.get.bind(blockService))(cid)`

vmx commented on src/index.js:
```diff
@@ -29,361 +25,450 @@ class IPLDResolver {
+    const codec = multicodec.getCode(codecBuffer)
+    if (this.resolvers[codec]) {
+      const codecName = multicodec.print[codec]
+      throw new Error(`Resolver already exists for codec "${codecName}"`)
```
I wonder how often you really want to replace an existing format. I see it as a gatekeeper to prevent hard to debug bugs, where accidentally some format is replaced.

vmx commented:
@alanshaw I think I addressed all your comments (btw: they were great, the code is much better now). I also added some information about the extended iterators that are returned by the API (see https://github.com/ipld/js-ipld/blob/f5fa4ceba7dc1f85ba771dce00b1679fc099d990/README.md#api).

alanshaw commented on src/index.js:
```diff
@@ -29,369 +24,338 @@ class IPLDResolver {
+    const serialized = await promisify(format.util.serialize)(node)
+    const block = new Block(serialized, cid)
+    await promisify(this.bs.put.bind(this.bs))(block)
+  }
```
I think this is ok for now but the block store needs to take a iterator and return an iterator that yields the blocks you put in:

```js
put (nodes) {
  return this.bs.put(serialize(nodes))
}
```

vmx commented on package.json:
```diff
@@ -33,36 +33,36 @@
-    "aegir": "^17.1.1",
-    "bitcoinjs-lib": "^4.0.2",
+    "aegir": "^18.0.3",
+    "async": "^2.6.1",
```
Yes for the tests. The tests should be refactored as there's a lot of duplicated code, but I'd do that at a later stage.

vmx commented on src/index.js:
```diff
@@ -1,21 +1,16 @@
 'use strict'
 
+const promisify = require('util').promisify
```
Could Node.js native things be polyfilled with something like promisify-es6?

vmx replied:
Nevermind, it seems to be compatible, so it was easy enough to switch.

vmx commented on src/index.js:
```diff
@@ -29,369 +24,338 @@ class IPLDResolver {
-    if (!Array.isArray(cids)) {
-      return callback(new Error('Argument must be an array of CIDs'))
+  get (cids) {
+    if (!typical.isIterable(cids) || typical.isString(cids) ||
```
Wow, good catch!

vmx commented on src/index.js:
```diff
@@ -29,369 +24,338 @@ class IPLDResolver {
+        const format = await this._getFormat(block.cid.codec)
+        const node = await promisify(format.util.deserialize)(block.data)
+        yield node
       }
```
Performance is not a concern for IPLD atm. It's not optimized at all yet.

vmx commented on src/index.js:
```diff
@@ -29,369 +24,338 @@ class IPLDResolver {
-            return e
+        // Only follow links if recursion is intended
+        if (options.recursive) {
+          cid = await maybeRecurse(block, treePath)
```
The `tree()` function will change a lot once the IPLD Formats use a new API. For now for me it's only important that it works.

vmx commented:
@alanshaw thanks for the thorough review. I've pushed new commits, all comments should be either addressed by those commits or my comments.

vmx commented:
My [PR on typical](https://github.com/75lb/typical/pull/2) to add async iterator support was merged an released!

achingbrain commented on src/index.js:
```diff
@@ -29,369 +23,342 @@ class IPLDResolver {
+        // get block
+        // use local resolver
+        // update path value
+        const block = await promisify(this.bs.get.bind(this.bs))(cid)
```
Why not do the promisification of the blockstore methods in the constructor instead of repeatedly during loops (e.g. here, during `.get`, `.remove`, `.tree`, etc)?

e.g.

```javascript
class IPLDResolver {
  constructor (opts) {
    this.bs = {
       get: promisify(opts.bs.get.bind(opts.bs)),
       put: promisify(opts.bs.put.bind(opts.bs)),
       delete: promisify(opts.bs.delete.bind(opts.bs)),
       // ...etc
    }

    // ...etc
  }
}
```

Alternatively use something like [Bluebird.promisifyAll](http://bluebirdjs.com/docs/api/promisification.html) until the lower level libraries are promise-aware?

vmx commented:
Pushed a new commit that adds the `*Many()` functions to match the most recent https://github.com/ipld/js-ipld/pull/191. @alanshaw it would be great if you could have a look.

rvagg commented on test/ipld-all.js:
```diff
@@ -52,57 +55,45 @@ describe('IPLD Resolver for dag-cbor + dag-pb', () => {
-
-        resolver.bs._repo.blocks.has(cid2, cb)
-      }
+      (node, cb) => {
```
this is looking awkward within a `waterfall`, maybe it should be converted to an async test?

vmx replied:
The tests were just adapted with minimal changes. The tests are in a bad shape currently and need a bigger refactor (there's lots of duplicated code). But I'd like to do it after this change is merged.

rvagg commented on test/ipld-dag-pb.js:
```diff
@@ -95,190 +96,104 @@ module.exports = (repo) => {
+        expect(sameAsNode.data).to.deep.equal(node.data)
+        return remove()
+
+        async function remove () {
```
why are these wrapped up in their own function? the two lines should probably just be at the top level of this test

vmx replied:
The tests were just adapted with minimal changes. The tests are in a bad shape currently and need a bigger refactor (there's lots of duplicated code). But I'd like to do it after this change is merged.

rvagg commented on src/index.js:
```diff
@@ -29,369 +23,394 @@ class IPLDResolver {
+   * @param {number} format - The multicodec of the format that IPLD Node should be encoded in.
+   * @param {Object} [userOptions] -  Options is an object with the following properties.
+   * @param {number} [userOtions.hashAlg=hash algorithm of the given multicodec] - The hashing algorithm that is used to calculate the CID.
+   * @param {number} [userOptions.cidVersion=1]`- The CID version to use.
```
erroneous '\`'

rvagg commented on src/index.js:
```diff
@@ -29,369 +23,394 @@ class IPLDResolver {
+   * @param {number} format - The multicodec of the format that IPLD Node should be encoded in.
+   * @param {Object} [userOptions] -  Options are applied to any of the `nodes` and is an object with the following properties.
+   * @param {number} [userOtions.hashAlg=hash algorithm of the given multicodec] - The hashing algorithm that is used to calculate the CID.
+   * @param {number} [userOptions.cidVersion=1]`- The CID version to use.
```
erroneous '\`'
